datafusion/functions/src/string/overlay.rs

```diff
@@ -88,10 +105,13 @@ pub fn overlay<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
             let characters_array = as_generic_string_array::<T>(&args[1])?;
             let pos_num = as_int64_array(&args[2])?;
 
+            let characters_array_iter = adaptive_array_iter(characters_array.iter());
+            let pos_num_iter = adaptive_array_iter(pos_num.iter());
+
             let result = string_array
                 .iter()
-                .zip(characters_array.iter())
-                .zip(pos_num.iter())
+                .zip(characters_array_iter)
+                .zip(pos_num_iter)
```

Comment:
While looking at the innter loop of this function, you can probably get a major boost by avoiding `String::with_capacoty` and instead reuse the same `String` on each element. It might also be possible to skip the copy by using StringBuilder directly rather than copying to a temp string